### DISCUSSION TITLE
3D measurement accuracy: has it been measured?; what can be expected?; how can it be made better?

### DISCUSSION THREAD

**Discussion by mjhmilla** [freemocap/freemocap#211] [General]
This is a delightful project and I'm really looking forward to trying it out! As I am in research, I have a round of accuracy questions:

1. Has the accuracy of the system been measured? 

2. What accuracy can be expected?
 a. Similar to what is reported by Nakano et al. with OpenPose and multiple cameras?
 b. Related: what algorithm is doing the pose recognition?

3. Are there ways to improve the accuracy? 
 a. What's the best accuracy that's ever been achieved? How was it achieved?
 
 Nakano, N., Sakura, T., Ueda, K., Omura, L., Kimura, A., Iino, Y., ... & Yoshioka, S. (2020). Evaluation of 3D markerless motion capture accuracy using OpenPose with multiple video cameras. Frontiers in sports and active living, 2, 50.

**Comment by aaroncherian** [comment]
Hey there - great questions. This is one paper that I was looking through a couple months back. 

1. So quantifying and measuring the accuracy of the system against a marker-based system is something we're working on now, and we're planning to look at it across a variety of different tasks. From a purely qualitative standpoint so far, what we're seeing looks promising. 

2. a. See above regarding expected accuracy. I think it would be hard to do a direct comparison between our system and the one used by Nakano et. all.  The tasks that they chose for comparison are a little different than the tasks that we're examining. They also only looked at mean accuracy error between certain joint positions, and while we will be looking at that for certain tasks, we also have other outcomes that we're calculating.  I will say that their methods are a bit opaque - so for example, their 3D reconstruction is based on methods written in a paper in 1980, so it's hard to spot where areas of inaccuracy might be in their pipeline. Ours is based on using more modern libraries in Python - which means that the methods are open for people to examine and that as we learn more we can improve and change our methods. So I'd say FreeMoCap can only continue to get more accurate as we work further on the software.
b. We're using Mediapipe, (https://google.github.io/mediapipe/), which is Google's pose detection algorithm. It's a lot more lightweight than OpenPose, so it makes for easier installation and faster runtime for our users. 

3. There are a lot of options for improving accuracy. Some are on the recording end - more cameras tends to give better results, making sure that the room is well lit, etc. There are also methods on the processing end that we can implement. Some of these we've made good progress in - there's one method that runs an object detection algorithm to identify and crop out the human in the frame, and then runs pose estimation on that cropped out image. So far this has seemed to increase the accuracy of our reconstruction while also preventing the pose estimation algorithm from misidentifying objects in the background as a human skeleton (which is also an issue Nakano et. all noted with their OpenPose estimations). There's a couple of other options that we might look into, such as gap filling or adding constraints, but those are more long-term goals of ours than things we're looking at in the short term. 
a. Again since we're not quantifying accuracy, I can't really give you a 'best accuracy' measurement - and I think that might also depend slightly on your use case for the software as well.

Hope this helps!

**Comment by jonmatthis** [comment]
Just to emphasize a few point in @aaroncherian's great response- 

- We (specificaly Aaron) are currently in the process of getting a better sense of the overall accuracy of the current system, but it's tricky because...

- ...it's often better think of 'accuracy' in terms of whether or not the data produced is sufficient *for a given use case.* This is doubly true for a project like this, where there are a whole laundry list of things that we could do to improve that accuracy. We haven't  been focusing on building out the core pipeline to improve usability and to ensure that the 'raw' data we produce is as good as we can get it, while keeping a lot of the options to clean up that raw data to the side while the core pipeline is still in development. Put another way, this is a long term project, and we fully expect our accuracy to get better over time as we improve our process (in addition to the core tracking algorithms improving external to this project)

So  if you are hoping to use `freemocap` for something on a short timescale, I'd encourage you to consider the specific use case you have in mind and then we can have a more directed conversation about whether our current system would be appropriate for that use, or if you should wait a few `%REASONABLE_TIME_UNITS%`  until we (or others? or you?) have implemented some of the potential post-processing clean up options for the raw data

**Reply by mjhmilla** [reply]
Wonderful response, thank you both @aaroncherian and @jonmatthis. Now that I know that the task-specific accuracy of the system is in progress, this is something I can help out with. This would make a great student project. Thank you again for taking the time to provide such detailed responses!

**Reply by jonmatthis** [reply]
Our pleasure! It's really rewarding to see people showing interest in this project, and we appreciate that you're willing to work with this thing when it's still-being-built form!

Assessing the feasibility/accuracy is definitely a great student project! In fact, I'm a professor myself, and Aaron is my grad student :) 

I'd invite you to point whichever student is working on this to come chat on the freemocap Discord server - https://discord.gg/pZW9R2fj

There's a growing community of users on there, and we're happy to help out and would love to hear what you're working on!

Thanks again for the interest! Welcome to the party :D
